<a href="https://colab.research.google.com/github/halimcan/Home-Credit-Default-Project/blob/installments_payments_branch6/Installments_payments.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Installments of required tables
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#  BigQuery API activation
from google.colab import auth
auth.authenticate_user()

In [3]:
# big query add-on installation

!pip install --quiet google-cloud-bigquery
from google.cloud import bigquery

In [9]:
# BigQuery client initiation
client = bigquery.Client(project="homecredit-478707")

# From BigQuery
query = """
SELECT *
FROM `homecredit-478707.Homecredit_Tables.Installments_payments`
"""
installments_payments = client.query(query).to_dataframe()

In [10]:
installments_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13605401 entries, 0 to 13605400
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   SK_ID_PREV              Int64  
 1   SK_ID_CURR              Int64  
 2   NUM_INSTALMENT_VERSION  float64
 3   NUM_INSTALMENT_NUMBER   Int64  
 4   DAYS_INSTALMENT         float64
 5   DAYS_ENTRY_PAYMENT      float64
 6   AMT_INSTALMENT          float64
 7   AMT_PAYMENT             float64
dtypes: Int64(3), float64(5)
memory usage: 869.3 MB


In [11]:
installments_payments.head(3)

SK_ID_PREV  SK_ID_CURR  NUM_INSTALMENT_VERSION  NUM_INSTALMENT_NUMBER  \
0     1949131      351869                     0.0                     58   
1     1944423      398027                     0.0                     64   
2     1029587      348709                     0.0                     88   

   DAYS_INSTALMENT  DAYS_ENTRY_PAYMENT  AMT_INSTALMENT  AMT_PAYMENT  
0          -1441.0             -1441.0           39.96        39.96  
1          -1259.0             -1260.0         9000.00      9000.00  
2           -860.0              -860.0          224.10       224.10

Why Installment Payments Cannot Use a Simple Primary Key

The installments_payments table contains one row per installment payment, but the dataset often includes multiple records for the same installment, caused by:

updates in the payment date

version changes in the repayment schedule

system-level duplicate entries

multiple recorded corrections or adjustments

In [12]:
duplicates = installments_payments.duplicated(
    subset=["SK_ID_CURR", "SK_ID_PREV", "NUM_INSTALMENT_NUMBER"]
).sum()

print("Duplicate PK rows:", duplicates)
print("Total rows:", installments_payments.shape[0])


Duplicate PK rows: 743407
Total rows: 13605401


In [13]:
duplicates_v2 = installments_payments.duplicated(
    subset=["SK_ID_CURR", "SK_ID_PREV", "NUM_INSTALMENT_NUMBER", "NUM_INSTALMENT_VERSION"]
).sum()

print("Duplicate PK rows with version:", duplicates_v2)


Duplicate PK rows with version: 653483


In [14]:
# Keep only the latest payment entry per installment
installments_payments_sorted = installments_payments.sort_values(
    by=["SK_ID_CURR", "SK_ID_PREV", "NUM_INSTALMENT_NUMBER", "NUM_INSTALMENT_VERSION", "DAYS_ENTRY_PAYMENT"],
    ascending=[True, True, True, True, True]
)

installments_unique = installments_payments_sorted.drop_duplicates(
    subset=["SK_ID_CURR", "SK_ID_PREV", "NUM_INSTALMENT_NUMBER"],
    keep="last"
)

print("Original rows:", installments_payments.shape[0])
print("After deduplication:", installments_unique.shape[0])


Original rows: 13605401
After deduplication: 12861994


How We Cleaned the Duplicates (Industry-Standard Approach)

To ensure a reliable and consistent installment-level dataset, we applied:

✔️ “Latest payment record” rule

For each installment, we kept the row with the most recent DAYS_ENTRY_PAYMENT, assuming it represents the most accurate and updated payment information.

This method:

eliminates duplicate payment logs

retains the latest corrected installment info

ensures a clean, one-row-per-installment structure

prepares the table for accurate behavior-based feature engineering

⭐ 2) Missing Value & Distribution — Do We Need Full EDA?
✔️ Short answer: No — not the classic EDA, and işte nedeni 👇

Installments table is transactional. Meaning:

It is not customer-level data

It is not application-level data

It is a time series of repayment actions

Therefore, missing values or distributions in the raw table do not have the same meaning as in application_train or previous_application.

⭐ Why missing values are NOT handled traditionally here
• DAYS_ENTRY_PAYMENT can be missing if the payment has not been made yet.

This is a behavioral signal, not a “missing value problem”.

• AMT_PAYMENT may be zero if the customer didn’t pay.

This is a payment default behavior, not something to impute.

• DAYS_INSTALMENT is always populated and not meant to be cleaned.
• NUM_INSTALMENT_VERSION may be duplicated for system reasons.

We already resolved this through deduplication.

Therefore:

❌ No median imputation
❌ No mode imputation
❌ No manual filling
❌ No column dropping

Because that would destroy behavioral patterns that models depend on.

⭐ 3) Why Distribution Plots Are Not Useful Here

For application_train:

histograms are meaningful

boxplots show real outliers

feature scaling matters

But for installments:

Installment values represent repeating monthly events.

Plotting distributions tells us nothing actionable because:

Each customer has a different loan structure

Every loan has 12–96+ installments

DAYS_INSTALMENT repeats in structured cycles

Amounts depend entirely on loan type, tenure, and schedule

📌 Distribution EDA is not only unhelpful, it’s actually misleading.

⭐ 4) What Matters Instead?

Installments tablosunda 1 numaralı amaç:

Extract behavioral signals

such as:

late payment count

early payment count

mean delay

overpayment ratio

underpayment frequency

max days late

payment discipline patterns

These FE features are far more meaningful than raw distributions.

Installments Table – Rationale Behind the Cleaning Approach

The installments table contains multiple recorded versions of the same installment due to payment updates and system-level replication.

A standard primary key does not exist; even the extended key including version fields is not unique.

To ensure consistent installment-level data, we kept the latest payment entry per installment, based on the maximum DAYS_ENTRY_PAYMENT value.

Missing values in this table reflect payment behavior (e.g., missed installments), not data quality problems, and therefore must not be imputed.

Distribution-based EDA is not meaningful for transactional time-series repayment data.

Instead, we engineered key behavioral features such as payment delay, underpayment and overpayment flags, payment ratios, and delinquency patterns.

These aggregated behavior-level signals significantly enhance the predictive power of the credit scoring model.

# Late / Early Payment FE

In [15]:
# Use deduplicated version
installments = installments_unique.copy()

# -----------------------------------------------
# 1) Days past due (late payment)
# PAST_DUE = DAYS_ENTRY_PAYMENT - DAYS_INSTALMENT
# -----------------------------------------------
installments["PAYMENT_DELAY"] = installments["DAYS_ENTRY_PAYMENT"] - installments["DAYS_INSTALMENT"]

# Late payment (positive delay)
installments["LATE_PAYMENT_FLAG"] = (installments["PAYMENT_DELAY"] > 0).astype(int)

# Early payment (negative delay)
installments["EARLY_PAYMENT_FLAG"] = (installments["PAYMENT_DELAY"] < 0).astype(int)

# On-time payment
installments["ON_TIME_PAYMENT_FLAG"] = (installments["PAYMENT_DELAY"] == 0).astype(int)


# ------------------------------------------------
# 2) Payment ratio
# PAYMENT_RATIO = AMT_PAYMENT / AMT_INSTALMENT
# ------------------------------------------------
installments["PAYMENT_RATIO"] = installments["AMT_PAYMENT"] / (installments["AMT_INSTALMENT"] + 1e-6)

# Underpayment (<1)
installments["UNDERPAYMENT_FLAG"] = (installments["PAYMENT_RATIO"] < 1).astype(int)

# Exact payment (=1)
installments["EXACT_PAYMENT_FLAG"] = (installments["PAYMENT_RATIO"].round(3) == 1).astype(int)

# Overpayment (>1)
installments["OVERPAYMENT_FLAG"] = (installments["PAYMENT_RATIO"] > 1).astype(int)


# ------------------------------------------------
# 3) Absolute delay magnitudes
# ------------------------------------------------
installments["LATE_DAYS"] = installments["PAYMENT_DELAY"].clip(lower=0)
installments["EARLY_DAYS"] = (-installments["PAYMENT_DELAY"]).clip(lower=0)


# Customer-Level Aggregation (SK_ID_CURR)

In [16]:
# Aggregation dictionary for installments
agg_dict = {
    "PAYMENT_DELAY": ["mean", "max", "min", "std", "median"],
    "LATE_DAYS": ["mean", "max"],
    "EARLY_DAYS": ["mean", "max"],
    "LATE_PAYMENT_FLAG": ["mean", "sum"],
    "EARLY_PAYMENT_FLAG": ["mean", "sum"],
    "ON_TIME_PAYMENT_FLAG": ["mean", "sum"],
    "PAYMENT_RATIO": ["mean", "min", "max", "std"],
    "UNDERPAYMENT_FLAG": ["mean", "sum"],
    "EXACT_PAYMENT_FLAG": ["mean", "sum"],
    "OVERPAYMENT_FLAG": ["mean", "sum"],
    "AMT_INSTALMENT": ["mean", "sum", "max"],
    "AMT_PAYMENT": ["mean", "sum", "max"]
}

installments_agg = installments.groupby("SK_ID_CURR").agg(agg_dict)

# flatten multi-index
installments_agg.columns = [
    f"INSTAL_{col[0]}_{col[1].upper()}" for col in installments_agg.columns
]

installments_agg.reset_index(inplace=True)

print("Final installments_agg shape:", installments_agg.shape)
installments_agg.head()


Final installments_agg shape: (339587, 32)


SK_ID_CURR  INSTAL_PAYMENT_DELAY_MEAN  INSTAL_PAYMENT_DELAY_MAX  \
0      100001                  -7.285714                      11.0   
1      100002                 -20.421053                     -12.0   
2      100003                  -7.160000                      -1.0   
3      100004                  -7.666667                      -3.0   
4      100005                 -23.555556                       1.0   

   INSTAL_PAYMENT_DELAY_MIN  INSTAL_PAYMENT_DELAY_STD  \
0                     -36.0                 14.625483   
1                     -31.0                  4.925171   
2                     -14.0                  3.726929   
3                     -11.0                  4.163332   
4                     -37.0                 13.510284   

   INSTAL_PAYMENT_DELAY_MEDIAN  INSTAL_LATE_DAYS_MEAN  INSTAL_LATE_DAYS_MAX  \
0                         -6.0               1.571429                  11.0   
1                        -19.0               0.000000                   0.0   
2                         -6.0               0.000000                   0.0   
3                         -9.0               0.000000                   0.0   
4                        -29.0               0.111111                   1.0   

   INSTAL_EARLY_DAYS_MEAN  INSTAL_EARLY_DAYS_MAX  ...  \
0                8.857143                   36.0  ...   
1               20.421053                   31.0  ...   
2                7.160000                   14.0  ...   
3                7.666667                   11.0  ...   
4               23.666667                   37.0  ...   

   INSTAL_EXACT_PAYMENT_FLAG_MEAN  INSTAL_EXACT_PAYMENT_FLAG_SUM  \
0                             1.0                              7   
1                             1.0                             19   
2                             1.0                             25   
3                             1.0                              3   
4                             1.0                              9   

   INSTAL_OVERPAYMENT_FLAG_MEAN  INSTAL_OVERPAYMENT_FLAG_SUM  \
0                           0.0                            0   
1                           0.0                            0   
2                           0.0                            0   
3                           0.0                            0   
4                           0.0                            0   

   INSTAL_AMT_INSTALMENT_MEAN  INSTAL_AMT_INSTALMENT_SUM  \
0                 5885.132143                  41195.925   
1                11559.247105                 219625.695   
2                64754.586000                1618864.650   
3                 7096.155000                  21288.465   
4                 6240.205000                  56161.845   

   INSTAL_AMT_INSTALMENT_MAX  INSTAL_AMT_PAYMENT_MEAN  INSTAL_AMT_PAYMENT_SUM  \
0                  17397.900              5885.132143               41195.925   
1                  53093.745             11559.247105              219625.695   
2                 560835.360             64754.586000             1618864.650   
3                  10573.965              7096.155000               21288.465   
4                  17656.245              6240.205000               56161.845   

   INSTAL_AMT_PAYMENT_MAX  
0               17397.900  
1               53093.745  
2              560835.360  
3               10573.965  
4               17656.245  

[5 rows x 32 columns]

Installments Aggregate – English Explanation (Presentation Ready)
What the aggregated table represents

The installments_agg table contains customer-level repayment behavior features derived from over 12.8 million installment payment records. Each row corresponds to one customer (SK_ID_CURR), and each column summarizes a specific type of credit behavior such as delay, early payment, underpayment, or overpayment.

* Why these features matter

Installment payment behavior is one of the strongest predictors of credit default.
The following patterns highly correlate with repayment risk:

Late payments → indicate financial stress

Early payments → indicate financial discipline or low risk

Underpayments → cash flow problems

Overpayments → high repayment capacity

Payment ratio variability → financial instability

Large maximum late days → high delinquency tendency

The 32 engineered features capture these behaviors at scale.

* How the aggregation works (in simple terms)

We transformed raw installment-level records into interpretable customer-level summaries:

* Delay-based features

INSTAL_PAYMENT_DELAY_MEAN

INSTAL_PAYMENT_DELAY_MAX

INSTAL_LATE_DAYS_MAX

These quantify how late or early customers tend to pay their installments.

* Payment precision features

INSTAL_UNDERPAYMENT_FLAG_MEAN

INSTAL_EXACT_PAYMENT_FLAG_MEAN

INSTAL_OVERPAYMENT_FLAG_MEAN

These indicate how often customers paid less, exactly, or more than required.

* Payment ratio & amount patterns

INSTAL_PAYMENT_RATIO_MEAN

INSTAL_PAYMENT_RATIO_STD

INSTAL_AMT_PAYMENT_MAX

INSTAL_AMT_INSTALMENT_SUM

* These measure the payment discipline and loan burden capacity.

*Final structure
Final installments_agg shape: (339,587 customers, 32 features)


This means:

We successfully compressed 12.8 million transactional records

Into 32 highly predictive features

At customer level, ready for merging into application_train

Why these FE features improve model performance

Because they encode actual historical payment behavior, which is more predictive than demographic or application-level data.

Models like LightGBM benefit massively from:

payment delay patterns

delinquency frequency

repayment consistency

stability of payment amounts

